### ISSUE TITLE
Hotswap not work on MacOS java version "1.8.0_112"

### ISSUE THREAD

**Issue by ulianakucheriava** [dcevm/dcevm#107]
HOTSWAP AGENT: 2:6:48.013 RELOAD (org.hotswap.agent.config.PluginManager) - Reloading classes [com.core.util.MccSiteUrlHelper](autoHotswap)
HOTSWAP AGENT: 2:6:48.013 RELOAD (org.hotswap.agent.config.PluginManager) - Reloading classes [com.core.util.MccSiteUrlHelper](autoHotswap)
HOTSWAP AGENT: 2:6:48.013 RELOAD (org.hotswap.agent.config.PluginManager) - Reloading classes [com.core.util.MccSiteUrlHelper](autoHotswap)
# 
# A fatal error has been detected by the Java Runtime Environment:
# 
# SIGSEGV (0xb) at pc=0x000000010fb21193, pid=1229, tid=0x0000000000002703
# 
# JRE version: Java(TM) SE Runtime Environment (8.0_112-b16) (build 1.8.0_112-b16)
# Java VM: Dynamic Code Evolution 64-Bit Server VM (25.71-b01-dcevmlight-dev mixed mode bsd-amd64 compressed oops)
# Problematic frame:
# V  [libjvm.dylib+0x17d193]
# 
# Failed to write core dump. Core dumps have been disabled. To enable core dumping, try "ulimit -c unlimited" before starting Java again
# 
# An error report file with more information is saved as:
# /Users/tomcat/bin/hs_err_pid1229.log
# 
# If you would like to submit a bug report, please visit:
# https://github.com/dcevm/dcevm/issues
# 

JVM received a signal UNKNOWN (6).
JVM process is gone.
JVM exited unexpectedly.

**Comment by skybber** [comment]
- Did you build your DCEVMu112 version? If not then don't use older decvm build(u92) in new openjdk(u112).
- Is it bug proproducible? 
- What kind of class change did you do before crash? (Method code modification, method add/remove or class field change or static field change, class addition, class inheritance modification) Some kind of modifications can cause DCEVM crash sometimes.
Anyhow send hs_err_pid1229.log within your bugreport, there are fundamental information about the crash.

**Comment by ulianakucheriava** [comment]
I built product version of DCEVM using hotspotTag=jdk8u112-b16.
Hotswap agent version [0.4.0-SNAPSHOT](https://github.com/HotswapProjects/HotswapAgent/releases/tag/0.4.0-SNAPSHOT%402016%2F10%2F12)
JVM crash logs: [https://gist.github.com/ulianakucheriava/bc98e3a0e794858b62e7ff3255e1987e](https://gist.github.com/ulianakucheriava/bc98e3a0e794858b62e7ff3255e1987e)
hs_err_pid.log: [https://gist.github.com/ulianakucheriava/2ffa682a06b0ade7b8ebd5532456c8c3](https://gist.github.com/ulianakucheriava/2ffa682a06b0ade7b8ebd5532456c8c3)

tomcat settings:

tomcat.debugjavaoptions=-Xdebug -Xnoagent -Xrunjdwp:transport=dt_socket,server=y,address=8000,suspend=n -javaagent:"/Users/zaelab/TDSC/HYBRISCOMM6200P_0-80002118/hybris/bin/platform/hotswap/hotswap-agent.jar"=autoHotswap=true,disablePlugin=Hibernate,disablePlugin=Jersey2,disablePlugin=Jetty,disablePlugin=ZK,disablePlugin=Logback,disablePlugin=Seam,disablePlugin=ELResolver,disablePlugin=OsgiEquinox,disablePlugin=WebObjects,disablePlugin=Weld,disablePlugin=JBossModules,disablePlugin=Resteasy,disablePlugin=Gae -XXaltjvm=dcevm

tomcat.generaloptions=-XX:+UseSerialGC -ea -Dcom.sun.management.jmxremote -Dcom.sun.management.jmxremote.authenticate=false -Dcom.sun.management.jmxremote.ssl=false -Dorg.tanukisoftware.wrapper.WrapperManager.mbean=true -Djava.endorsed.dirs="%CATALINA_HOME%/lib/endorsed" -Dcatalina.base=%CATALINA_BASE% -Dcatalina.home=%CATALINA_HOME% -Dfile.encoding=UTF-8 -Dlog4j.configuration=log4j_init_tomcat.properties -Djava.util.logging.config.file=jdk_logging.properties -Djava.io.tmpdir="${HYBRIS_TEMP_DIR}"

it also crashes without using tomcat.generaloptions

**Comment by skybber** [comment]
I this valid Java syntax ?

     String LOG.error("TEST");

Does it work with DCEVMu92?

**Comment by ulianakucheriava** [comment]
Sorry, wrong copypaste. I just add LOG.error("TEST"); to method. I also try to add System.out.println("Test"); and it still crashes. Java syntax is valid, classes builds succesfully, but dcevm crashes.

**Comment by skybber** [comment]
According stack trace there are 2 threads calling to JvmtiEnv::RedefineClasses. There used to be some locking mechanism in dcevm7... Is this problem permanent or occasional?

**Comment by skybber** [comment]
I've checked your settings one again and the problem is obvious. You are running remote debugging and autoHotswap=true simultaneously. You should choose either autoHotswap=true or jdwp since using both of them can lead to parallel request to jvm-ti redefinition. Dcevm should lock the redefinition method but it doesn't do it now.

**Comment by oranheim** [comment]
@ulianakucheriava: Are you able to build DCEVM for OpenJDK8-u112-b16 on MacOS Sierra? It would be of great valuable if you could share your build steps.. 🥇  I followed the docs, but it didn't work out. There is another issue related to this: https://github.com/dcevm/dcevm/issues/111.

I have installed `brew gcc-48` and set my env vars:

```
export MACOSX_DEPLOYMENT_TARGET=10.8
export CC=/usr/local/bin/gcc-4.8
export CFLAGS=-fobjc-exceptions
export CXX=/usr/local/bin/g++-4.8
export SA_LDFLAGS=-fobjc-exceptions
```

then `./gradlew patch && ./gradlew ./gradlew compileFastdebug`

I get tons of linking errors, like:

```
13:55:24.946 [INFO] [system.out] Linking vm...
13:55:25.429 [ERROR] [system.err] In file included from /usr/local/Cellar/gcc@4.8/4.8.5/lib/gcc/4.8/gcc/x86_64-apple-darwin16.3.0/4.8.5/include-fixed/Availability.h:205:0,
13:55:25.429 [ERROR] [system.err]                  from /usr/local/Cellar/gcc@4.8/4.8.5/lib/gcc/4.8/gcc/x86_64-apple-darwin16.3.0/4.8.5/include-fixed/objc/objc.h:47,
13:55:25.429 [ERROR] [system.err]                  from /usr/include/objc/runtime.h:27,
13:55:25.429 [ERROR] [system.err]                  from /usr/include/objc/objc-runtime.h:1,
13:55:25.429 [ERROR] [system.err]                  from /Users/oranheim/IdeaProjects/dcevm/hotspot/agent/src/os/bsd/MacosxDebuggerLocal.m:25:
13:55:25.429 [ERROR] [system.err] /System/Library/Frameworks/CoreFoundation.framework/Headers/CFDateFormatter.h:53:34: error: expected ',' or '}' before '__attribute__'
13:55:25.429 [ERROR] [system.err]      kCFISO8601DateFormatWithYear API_AVAILABLE(macosx(10.12), ios(10.0), watchos(3.0), tvos(10.0)) = (1UL << 0),
...(many lines deleted)...
```

I can't figure it out.

**Comment by skybber** [comment]
Fixed by lock on jvmti-redefine.
